Add --create option to `flutter emulators` command

DanTup opened the pull request:
This is still a bit rough and has no tests, but I wanted to nail down the logic before I start adding tests against it. This is based on the logic described in https://github.com/flutter/flutter/issues/13379#issuecomment-394686208. I don't entirely like how we have to do it (for ex. trying to create on without the sdk ID because we can't get a list of these, only devices/targets) but (on my Macbook at least) it seems to work.

LMK what you think. We may also wish to review the device we pick (currently we try `pixel` and `pixel_xl`) and whether we want to lock to a specific Android API version or just take the highest (which is what this currently does).

<img width="661" alt="screen shot 2018-06-06 at 1 13 35 pm" src="https://user-images.githubusercontent.com/1078012/41037511-75e1a718-698b-11e8-8e91-180b82f6da34.png">

(cc @devoncarew @mit-mit)

DanTup committed:
Add --create option to flutter emulators

devoncarew commented on packages/flutter_tools/lib/src/android/android_sdk.dart:
```diff
@@ -104,6 +104,23 @@ String getAvdPath() {
+/// This should be used over accessing androidSdk.avdManagerPath directly because it
+/// will work for those users who have Android Tools installed but
+/// not the full SDK.
+String getAvdManagerPath([AndroidSdk existingSdk]) {
```
use a named param instead?

DanTup replied:
This matches other functions in this file. Should we change them to keep them consistent?

```dart
String getAdbPath([AndroidSdk existingSdk]);
String getEmulatorPath([AndroidSdk existingSdk]);
```

devoncarew replied:
No, let's keep them consistent.

devoncarew commented on packages/flutter_tools/lib/src/android/android_sdk.dart:
```diff
@@ -104,6 +104,23 @@ String getAvdPath() {
+
+  final AndroidSdk sdk = AndroidSdk.locateAndroidSdk();
+
+  if (sdk?.latestVersion == null) {
```
I suspect this fallback logic parallels other android tools lookups in the source base (adb?). We may want to just fail if we can't locate an android sdk however (not do the path based lookup for `avdmanager`).

DanTup replied:
Currently there are three possibilities in here (same for the other few in this file):

1. If it exists on the SDK passed in (which looks to come from `context[AndroidSdk]`), we use that
2. If we can locate an SDK, that's used
3. Otherwise we try `os.which('emulator')?.path`

I'm not sure which out 1 and 2 is the best one (or whether we need 3). WDYT? If going with 1 then we can just delete this function; but I'm not sure whether that's a good bet. There are some other paths that do seem to use that and bail out if it's not set without any fallback, eg.:

```dart
  Future<bool> _checkForSupportedAdbVersion() async {
    if (androidSdk == null)
      return false;
```

Though interestingly, it then calls `getAdbPath` which does the fallback; but in this case it would never trigger because of the `return` above.

devoncarew replied:
Let's just delete the `os.which('emulator')?.path` option.

> Though interestingly, it then calls getAdbPath which does the fallback; but in this case it would never trigger because of the return above.

Hmm, sounds like some logic was orphaned by a refactor. Falling back to 'which `adb`' was useful since the user could then have a little functionality - like listing devices - even if they didn't have the full functionality that configuring an android sdk would give. If this regressed at some point, and people didn't notice, than it's likely that we don't need the fallback even for `adb`.

DanTup replied:
Ok, I'll remove the `which`. Looks like there are still some code paths that don't do what I described above, including `getAdbDevices`, so my comment about was a little misleading. Whether or not `_checkForSupportedAdbVersion` is behaving as we'd like though, I'm not sure. LMK if you think this needs some investigation and I'll take a look outside of this PR.

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -16,6 +16,8 @@ class EmulatorsCommand extends FlutterCommand {
   EmulatorsCommand() {
     argParser.addOption('launch',
         help: 'The full or partial ID of the emulator to launch.');
+    argParser.addOption('create',
```
Should `--create` be a flag (instead of an option)? We could auto-name the resulting image (`pixel`, `pixel_xl-2`, ...), and have an optional flag to name a created image.

devoncarew replied:
`create` feels like a subcommand, even if the `package:args` api doesn't make it easy to have it as one.

DanTup replied:
> Should `--create be a flag` (instead of an option)?  We could auto-name the resulting image

I was a bit unsure about this. If we pick the name for the user then it feels weird if we tell them "an emulator with that name already exists" so I was going to make them always provide it.

Of course, we could do the check ourselves first and only prompt if it already exists, but that might be better handled in the editor than here so it has better control over how to present it.

I was wondering if we should expose an `emulator.getDeviceProfiles`. That way when the user chooses "create emulator" in the editor, we can fetch the possible profiles (and filter it to pixels, nexuses, whatever) and then let them pick a device. I think being able to pick between a phone and a tablet could be helpful?

devoncarew replied:
> If we pick the name for the user then it feels weird if we tell them "an emulator with that name already exists"

We could choose a name, see if it already exists, and suffix something to de-dup (my_name-2).

> and then let them pick a device

I generally like to avoid asking the user for input on stdin. I think `flutter emulators --create --name=my_device` would be a reasonable UI, but am not fixed on it if you think there's a better way.

DanTup replied:
Ok, gone with this. We only add suffixes if we're generating the name, so if the user provides an explicit name we'll just give them an error:

```
dantup-macbookpro:flutter dantup$ flutter emulator --create
Emulator 'flutter_emulator' created successfully.

dantup-macbookpro:flutter dantup$ flutter emulator --create
Emulator 'flutter_emulator_1' created successfully.

dantup-macbookpro:flutter dantup$ flutter emulator --create --name flutter_emulator_1
Failed to create emulator 'flutter_emulator_1'.
Error: Android Virtual Device 'flutter_emulator_1' already exists.
```

It feels a bit weird doing `flutter emulator --create` and not `flutter emulator create` but I think that's unavoidable as-is.

Do we have any convention for indicating when a flag is only usable in conjunction with another (eg. `--name` only makes sense with `--create`)? I couldn't see anything obvious.

devoncarew replied:
> Do we have any convention for indicating when a flag is only usable in conjunction with another (eg. --name only makes sense with --create)? I couldn't see anything obvious

Not that I know of; likely just something we should do w/ the help text for the flags.

DanTup replied:
I found some are prefixed with `Used with flag --xxx.` so I'll copy that.

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -65,6 +69,18 @@ class EmulatorsCommand extends FlutterCommand {
+      printStatus("Emulator '$name' created successfully.");
+    } else {
+      printStatus(createResult.error);
+      printStatus("Failed to create emulator '$name'.");
```
We should print some text here about where they can find more info about creating emulator images (referencing flutter.io, or Android / Android Studio docs).

DanTup replied:
How's this? Seems a bit weird sending people to Android Studio or avdmanager docs but since the errors come from them I'm not sure if we can provide the info ourselves.

<img width="523" alt="screen shot 2018-06-11 at 6 00 30 pm" src="https://user-images.githubusercontent.com/1078012/41245950-6df61d1c-6da1-11e8-8f4d-df2be22c9d7d.png">

devoncarew replied:
This looks great for error text. We'll probably also want something brief for `flutter emulators -h` help text as well - perhaps pointing to a flutter.io page?

@mit-mit, @mjohnsullivan - we may want to consider a brief landing page on flutter.io describing how to create android emulators, or at least just routing people to places to find for info about it.

devoncarew reviewed (commented):
Thanks for starting this! I added some comments on the CLI UI.

DanTup committed:
Tweaks to error message

DanTup committed:
Simplify emulator search logic

DanTup committed:
Make name optional

devoncarew commented:
> This still badly needs some tests - assuming the general idea is ok I'll work on them this week.

Idea looks great!

DanTup committed:
Add a note about this option being used with --create

DanTup committed:
Tweaks to help information

DanTup committed:
Switch to processManager for easier testing

DanTup committed:
Don't crash on missing files or missing properties in Android Emulator

DanTup committed:
Move name suffixing into emulator manager

This allows it to be tested in the EmulatorManager tests and also used by daemon later if desired.

DanTup committed:
Pass the context's android SDK through so it can be mocked by tests

DanTup committed:
Misc fixes

DanTup committed:
Add tests around emulator creation

Process calls are mocked to avoid needing a real SDK (and to be fast). Full integration tests may be useful, but may require ensuring all build environments etc. are set up correctly.

DanTup commented:
Ok, I've pushed some tests for this. It turned out to be a bit more involved than I expected - I had to make some changes to the existing code to use processManager (so it can be mocked) plus some other tweaks. I tried to split the commits up to make it easier to understand all the changes.

I don't know if this is all the best way to do this (I've not used Mockito before and I'm not sure of our preferences between mocked and real integration tests) but this basically mocks the process manager to return canned responses instead of calling real SDK commands (I wasn't sure how well that's work in the build environment). Please review carefully and nitpick away!

I'm going to do some additional manual testing on Windows shortly, but seemed worthwhile kicking the CI run off to get the results quicker.

DanTup committed:
Simplify avdManagerPath

Previous changes were to emulatorPath!

DanTup committed:
Fix lint errors

DanTup committed:
Fix incorrect file exgtension for Windows

DanTup committed:
Fix an issue where no system images would crash

reduce throws on an empty collection.

DanTup committed:
Fix "null" appearing in error messages

The name we attempted to use will now always be returned, even in the case of failure.

DanTup committed:
Add additional info to missing-system-image failure message

On Windows after installing Andriod Studio I didn't have any of these and got this message. Installing with sdkmanager fixed the issue.

DanTup committed:
Fix thrown errors

runResult had a toString() but we moved to ProcessResult when switching to ProcessManager to this ended up throwing "Instance of ProcessResult".

DanTup committed:
Fix package import

DanTup committed:
Fix more package imports

DanTup committed:
Move mock implementation into Mock class

There seemed to be issues using Lists in args with Mockito that I couldn't figure out (docs say to use typed() but I couldn't make this compile with these lists still)..

DanTup committed:
Rename method that's ambigious now we have create

DanTup committed:
Handle where there's no avd path

DanTup commented:
Apparently this code had many issues! :(

I've fixed a bunch of bugs from testing on Windows and on the build servers (eg. due to the avd folder didn't exist) and improved some errors. We're touching a lot of stuff so I suspect we may find more edge cases at it runs in more places.

After this, I think it'd might be good to have a real end-to-end test that does something like:

- create emulator
- launch emulator
- check a new device appeared in `flutter devices`
- maybe even launch an app on it?

I'm not sure how fragile that might be (or how well the build infrastructure will support it - I'd guess the devicelab machines should be able to do it?) but it might be a good check.

DanTup commented:
I seem to get a lot of issues like this on build servers:

```
00:18 +109: test/emulator_test.dart: EmulatorManager create emulator with an empty name does not fail                                                                                                  type 'MappedListIterable' is not a subtype of type 'List<int>' of 'availableApiVersions'
```

Yet get no warnings locally, even with `--preview-dart-2` passed to VM and analyzer. Not sure if I'm doing something wrong?

DanTup committed:
Add another toList() :(

DanTup committed:
Remove comment that was rewritten

DanTup committed:
Fix forbidden import

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -70,17 +77,27 @@ class EmulatorsCommand extends FlutterCommand {
     }
   }
 
+  Future<Null> _createEmulator(String name) async {
```
Shouldn't `name` be optional here?

devoncarew replied:
Cool, I think it just makes it a bit more explicit that it can be null (in the absence of non-null annotations).

devoncarew commented on packages/flutter_tools/lib/src/emulator.dart:
```diff
@@ -57,6 +60,136 @@ class EmulatorManager {
+    
+    final String device =  await _getPreferredAvailableDevice();
+    if (device == null)
+      return new CreateEmulatorResult(
```
nit: the formatting here looks off. 2 space indent? no trailing comma?

Since this is a newer file - added by you recently - you may just want to dartfmt it as a level-set.

DanTup replied:
Done (though I tweaked some of its changes where they seemed to make it less readable).

devoncarew commented on packages/flutter_tools/lib/src/emulator.dart:
```diff
@@ -57,6 +60,136 @@ class EmulatorManager {
+        error: 'No device definitions are available'
+    );
+
+    final String sdkId =  await _getPreferredSdkId();
```
nit: two spaces before await.

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -16,13 +16,18 @@ class EmulatorsCommand extends FlutterCommand {
     argParser.addOption('launch',
         help: 'The full or partial ID of the emulator to launch.');
+    argParser.addFlag('create',
+        help: 'Creates a new Android emulator.',
```
We may want to make this a bit longer; we could say that we'll try and create a pixel device, and perhaps include a goo.gl shortened link for more info (until we have a short link to a flutter.io page).

DanTup replied:
I added "based on a Pixel device" though not sure what more to add. Do we have somewhere reasonable to link to? The current links we have (in the help) go to pages on managing AVDs with the Android SDK which might not be totally appropriate here?

DanTup committed:
Make optional arg more obviously optional

DanTup committed:
Reformat doc

DanTup committed:
Note that we create a pixel device in help text

devoncarew commented on packages/flutter_tools/lib/src/commands/emulators.dart:
```diff
@@ -77,7 +77,7 @@ class EmulatorsCommand extends FlutterCommand {
   }
 
-  Future<Null> _createEmulator(String name) async {
+  Future<Null> _createEmulator([String name]) async {
```
As a nit, I generally use named params.

DanTup committed:
Make this a named arg
